heaps/min_heap.py

```diff
@@ -11,73 +14,123 @@ def __repr__(self):
+            Time Complexity: O(log(n)) : remove head and replace with tail is constant. re balancing is log(n)
+            Space Complexity: O(1) :
         """
-        pass
+        if self.empty():
+            return
+        min = self.store[0]  # set aside root value
+        self.store[0] = self.store[-1]  # Delete root value and replace with last node
+        self.store.pop(-1)
+        self.heap_down(0)  # re balance the tree from the root position
+        return min[1]  # return value removed from heap
 
 
-    
     def __str__(self):
         """ This method lets you print the heap, when you're testing your app.
         """
         if len(self.store) == 0:
             return "[]"
-        return f"[{', '.join([str(element) for element in self.store])}]"
+        return f"[{', '.join([str(element[1]) for element in self.store])}]"
 
 
     def empty(self):
         """ This method returns true if the heap is empty
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: O(nlogn)
+            Space complexity: O(n)
         """
-        pass
+        if len(self.store) == 0:
+            return True
+
+        return False  # ?? implicit?
 
 
+    # "Bubble up" to root
     def heap_up(self, index):
         """ This helper method takes an index and
-            moves the corresponding element up the heap, if 
+            moves the corresponding element up the heap, if
             it is less than it's parent node until the Heap
             property is reestablished.
-            
+
             This could be **very** helpful for the add method.
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: O(logn)
+            Space complexity: O(n +1)
         """
-        pass
+        if index == 0:
+            return
+
+        if index == 1 or index == 2:
+            parent_index = 0
+        else:
+            parent_index = int(math.ceil((index-2)/2))
+            # if parent_index <= 1:
+            #     parent_index = int(1)
+
+        print("Parent")
+        print(parent_index)
+        print(type(parent_index))
+        print("index")
+        print(index)
+
+
+        if self.store[int(parent_index)][0] > self.store[index][0]:
+            self.swap(index, parent_index)
+            self.heap_up(parent_index)
+
+        return
 
+    # "filter down" to child position.
     def heap_down(self, index):
-        """ This helper method takes an index and 
-            moves the corresponding element down the heap if it's 
+        """ This helper method takes an index and
+            moves the corresponding element down the heap if it's
             larger than either of its children and continues until
             the heap property is reestablished.
         """
-        pass
+        left_index = (index*2)+1
+        right_index = (index*2)+2
+
+        if left_index >= len(self.store) or right_index >= len(self.store):
```

Comment:
You can have a left child with no right child. 

```suggestion
        if left_index >= len(self.store) and right_index >= len(self.store):
```